Source/DafnyCore/DafnyFile.cs

```diff
@@ -14,63 +14,43 @@ public class DafnyFile {
   public string BaseName { get; private set; }
   public bool IsPreverified { get; set; }
   public bool IsPrecompiled { get; set; }
-  public string SourceFilePath { get; private set; }
-  public Uri Uri { get; private set; }
-
-  // Returns a canonical string for the given file path, namely one which is the same
-  // for all paths to a given file and different otherwise. The best we can do is to
-  // make the path absolute -- detecting case and canonicalizing symbolic and hard
-  // links are difficult across file systems (which may mount parts of other filesystems,
-  // with different characteristics) and is not supported by .Net libraries
-  public static Uri Canonicalize(string filePath) {
-    if (filePath == null || !filePath.StartsWith("file:")) {
-      return new Uri(Path.GetFullPath(filePath));
-    }
-
-    if (Uri.IsWellFormedUriString(filePath, UriKind.RelativeOrAbsolute)) {
-      return new Uri(filePath);
-    }
-
-    var potentialPrefixes = new List<string>() { "file:\\", "file:/", "file:" };
-    foreach (var potentialPrefix in potentialPrefixes) {
-      if (filePath.StartsWith(potentialPrefix)) {
-        var withoutPrefix = filePath.Substring(potentialPrefix.Length);
-        var tentativeURI = "file:///" + withoutPrefix.Replace("\\", "/");
-        if (Uri.IsWellFormedUriString(tentativeURI, UriKind.RelativeOrAbsolute)) {
-          return new Uri(tentativeURI);
-        }
-        // Recovery mechanisms for the language server
-        return new Uri(filePath.Substring(potentialPrefix.Length));
-      }
-    }
-    return new Uri(filePath.Substring("file:".Length));
-  }
-  public static List<string> FileNames(IList<DafnyFile> dafnyFiles) {
-    var sourceFiles = new List<string>();
-    foreach (DafnyFile f in dafnyFiles) {
-      sourceFiles.Add(f.FilePath);
-    }
-    return sourceFiles;
-  }
-  public DafnyFile(DafnyOptions options, string filePath, bool useStdin = false) {
-    UseStdin = useStdin;
-    Uri = useStdin ? new Uri("stdin:///") : new Uri(filePath);
-    BaseName = useStdin ? "<stdin>" : Path.GetFileName(filePath);
-
-    var extension = useStdin ? ".dfy" : Path.GetExtension(filePath);
+  public TextReader Content { get; }
+  public Uri Uri { get; }
+  public DafnyFile(DafnyOptions options, string filePath, TextReader contentOverride = null) {
```

Comment:
nit: It'd be good to name `contentOverride` to suggest that it should only be used for stdin (unless that's not true, but it seems so based on the implementation)